src/tools/erasure-code/ceph-erasure-code-tool.cc

```diff
@@ -0,0 +1,174 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "include/buffer.h"
+#include "include/stringify.h"
+#include "common/ceph_argparse.h"
+#include "common/config_proxy.h"
+#include "common/errno.h"
+#include "erasure-code/ErasureCode.h"
+#include "erasure-code/ErasureCodePlugin.h"
+#include "global/global_context.h"
+#include "global/global_init.h"
+#include "osd/ECUtil.h"
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include <boost/algorithm/string.hpp>
+
+void usage(const std::string message, ostream &out) {
```

Comment:
instead of creating a new tool, Mykola, have you considered adding this feature to "ceph_erasure_code"?

Reply:
"ceph_erasure_code" is in tests. If we decide my `ceph-erasure-code-tool` should be in the test package then I think it makes sense and I will combine both.

Reply:
no. i don't think we should have this tool in the test package. how about moving ceph_erasure_code to ceph-osd and add this feature to ceph_erasure_code ?

Reply:
By "moving ceph_erasure_code to ceph-osd" I suppose you mean adding it to ceph-osd package?

What about src location? Should I move it from test to tool (where my ceph-erasure-code-tool is currently located)? And I am not sure about the name (ceph_erasure_code). I don't know if we have some rules about this, but looking at the current binary names, the tools use hyphens in their names, while the test binaries use underscores. And usually tools have "-tool" suffix in their names. That is why I named it ceph-erasure-code-tool. Making it `ceph_erasure_code` seems to be not following the convention.

So I vote either to leave it as it is now and update ceph.spec.in and debian/rules to add `ceph-erasure-code-tool` to ceph-osd, or add encode/decode functionality to ceph_erasure_code test binary and make sure it is packaged in ceph-test.

I prefer the first option but fine with the second one too.

As for the first option, additionally I could add "test" (or "test-profile", "check-profile"?) command, that would provide functionality that `ceph_erasure_code` currently provides and we could probably remove `ceph_erasure_code`. I also can consider using options like `--parameter` (which `ceph_erasure_code` uses) to specify EC profile instead of positional arguments, as in my current version, if people like this more.

Reply:
> By "moving ceph_erasure_code to ceph-osd" I suppose you mean adding it to ceph-osd package?

yes. i should have been more explicit. (i just edited the reply to clarify this).

> 
> What about src location? Should I move it from test to tool (where my ceph-erasure-code-tool is currently located)?

i think so if no objections.

> And I am not sure about the name (ceph_erasure_code). I don't know if we have some rules about this, but looking at the current binary names, the tools use hyphens in their names, while the test binaries use underscores. And usually tools have "-tool" suffix in their names. That is why I named it ceph-erasure-code-tool. Making it `ceph_erasure_code` seems to be not following the convention.

yeah, so far, i propose

1. move ceph_erasure_code to ceph-osd package
2. rename it to ceph_erasure_code ceph-erasure-code-tool
3. add the denc feature to ceph-erasure-code-tool

> 
> So I vote either to leave it as it is now and update ceph.spec.in and debian/rules to add `ceph-erasure-code-tool` to ceph-osd, or add encode/decode functionality to ceph_erasure_code test binary and make sure it is packaged in ceph-test.
> 
> I prefer the first option but fine with the second one too.

i still suggest combine these two tools into one. i can help on the packaging things if that worries you.

@neha-ojha @jdurgin what do you think?

> 
> As for the first option, additionally I could add "test" (or "test-profile", "check-profile"?) command, that would provide functionality that `ceph_erasure_code` currently provides and we could probably remove `ceph_erasure_code`. I also can consider using options like `--parameter` (which `ceph_erasure_code` uses) to specify EC profile instead of positional arguments, as in my current version, if people like this more.

Reply:
Combining both into ceph-erasure-code-tool in the ceph-osd package, with the existing profile functionality as a subcommand as @trociny suggested sounds good to me.

BTW, this is a great addition and will be incredibly useful for handling delicate EC situations. Just a couple weeks ago I was thinking we should have a tool like this!

Reply:
+1